ext/spl/spl_fixedarray.c

```diff
@@ -389,10 +389,14 @@ static inline void spl_fixedarray_object_write_dimension_helper(spl_fixedarray_o
 		zend_throw_exception(spl_ce_RuntimeException, "Index invalid or out of range", 0);
 		return;
 	} else {
-		if (!Z_ISUNDEF(intern->array.elements[index])) {
-			zval_ptr_dtor(&(intern->array.elements[index]));
+		/* Fix #81429 */
+		zval *const ptr = &(intern->array.elements[index]);
+		zval tmp;
+		ZVAL_COPY_VALUE(&tmp, ptr);
+		ZVAL_COPY_DEREF(ptr, value);
+		if (!Z_ISUNDEF(tmp)) {
+			zval_ptr_dtor(&tmp);
```

Comment:
`zend_assign_to_variable(ptr, value, IS_VAR, /* strict */ 0)`?

Reply:
I think that's the wrong abstraction - when I promote this change from PHP 7.4 to PHP 8.0, I'll no longer need to handle IS_UNDEF due to ca8657a2b502791b1ec0587948741098ca82e314 `Initialize SplFixedArray elements to NULL instead of UNDEF`

And SplFixedArray can't contain references (always dereferencing before assigning to entries), so zend_assign_to_variable seems like the wrong abstraction

So the additional code in the implementation of zend_assign_to_variable makes it less efficient, and less clear what this is trying to do

Also, I definitely don't want to use `Z_OBJ_HANDLER_P(variable_ptr, set)` here, that would be a bug.

```php
	do {
		if (UNEXPECTED(Z_REFCOUNTED_P(variable_ptr))) {
			zend_refcounted *garbage;

			if (Z_ISREF_P(variable_ptr)) {
				if (UNEXPECTED(ZEND_REF_HAS_TYPE_SOURCES(Z_REF_P(variable_ptr)))) {
					return zend_assign_to_typed_ref(variable_ptr, value, value_type, strict, ref);
				}

				variable_ptr = Z_REFVAL_P(variable_ptr);
				if (EXPECTED(!Z_REFCOUNTED_P(variable_ptr))) {
					break;
				}
			}
			if (Z_TYPE_P(variable_ptr) == IS_OBJECT &&
	    		UNEXPECTED(Z_OBJ_HANDLER_P(variable_ptr, set) != NULL)) {
				Z_OBJ_HANDLER_P(variable_ptr, set)(variable_ptr, value);
				return variable_ptr;
			}
			garbage = Z_COUNTED_P(variable_ptr);
			zend_copy_to_variable(variable_ptr, value, value_type, ref);
			if (GC_DELREF(garbage) == 0) {
				rc_dtor_func(garbage);
			} else { /* we need to split */
				/* optimized version of GC_ZVAL_CHECK_POSSIBLE_ROOT(variable_ptr) */
				if (UNEXPECTED(GC_MAY_LEAK(garbage))) {
					gc_possible_root(garbage);
				}
			}
			return variable_ptr;
		}
	} while (0);
```

Reply:
Efficiently dealing with the dtor issue is the primary purpose of zend_assign_to_variable, but I agree that it is overkill here, as we're in a more constrained situation.